### PR TITLE
feat: Entity twin data CRUD REST APIs

### PR DESCRIPTION
**Pull request by albinsuresh**
## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s. You can activate automatic signing by running `just prepare-dev` once)
- [ ] I ran `just format` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `just check` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge_agent/src/http_server/entity_store.rs [hunk 1]
```diff
@@ -170,13 +181,113 @@ async fn register_entity(
...
+        [seg1, seg2, seg3, seg4, "twin", fragment_key] => {
+            let topic_id = topic_id_from_path_segments(seg1, Some(seg2), Some(seg3), Some(seg4))?;
+            Ok((
+                topic_id,
+                Channel::EntityTwinData {
+                    fragment_key: fragment_key.to_string(),
+                },
+            ))
+        }
+        _ => Err(Error::PathNotFound(path.to_string())),
```
Comment (didier-wenzek) [inline crates/core/tedge_agent/src/http_server/entity_store.rs:266]: The error cases must give move insight to the user.

Currently, the followings lead to the same error message telling the path is not found.

* Incorrect channel: `device/main///cmd/maintenance_window`
* Missing fragment key: `device/main///twin`
* Incorrect fragment key: `device/main///twin/maintenance/window not`

**Review by didier-wenzek — COMMENTED** [review]
The code is correct but the error messages have to be improved.

File: crates/core/tedge_agent/src/http_server/entity_store.rs [hunk 2]
```diff
@@ -143,9 +153,10 @@ pub(crate) fn entity_store_router(state: AgentState) -> Router {
         .route("/v1/entities", post(register_entity).get(list_entities))
         .route(
             "/v1/entities/{*path}",
-            get(get_entity)
-                .patch(patch_entity)
-                .delete(deregister_entity),
+            get(get_resource)
+                .put(put_resource)
+                .patch(patch_resource)
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/core/tedge_agent/src/http_server/entity_store.rs:158]: > It's not really clear how patch is intended to be used.

The `PATCH` method is only implemented for `entities/<entity-topic-id>` endpoint, where the resource is the entity itself, as patch here means you're patching the existing entity with additional twin data. But, `PATCH` is not implemented for the `twin/<fragment_key>` endpoint, where the resource is the twin fragment itself, as inserts/updates of the twin values were conceived as `PUT` instead of a `PATCH`. I didn't overload `PATCH` here to keep that option open for patching an existing twin fragments by adding/updating sub-fragments.

> But the variant above gives an error message that is difficult to understand

Yeah, that's misleading. It should have been a 405 Method Not Allowed.

File: crates/core/tedge_agent/src/http_server/entity_store.rs (hunk 1 above)
Comment (albinsuresh) [reply crates/core/tedge_agent/src/http_server/entity_store.rs:266]: I was just trying to mimic the simple `Not Found` message emitted by axum when a non-existent resource path is queried. Will fix that by adding more context.

File: crates/core/tedge_agent/src/http_server/entity_store.rs [hunk 3]
```diff
@@ -115,6 +117,12 @@ enum Error {
 
     #[error(transparent)]
     InvalidTwinData(#[from] InvalidTwinData),
+
+    #[error("Path: {0} not found")]
+    PathNotFound(String),
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/core/tedge_agent/src/http_server/entity_store.rs:122]: I was just trying to mimic the simple Not Found message emitted by axum when a non-existent resource path is queried. Will fix that by adding more error variants.

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot [hunk 4]
```diff
@@ -126,6 +127,28 @@ Delete entity tree
     ...    {"@topic-id":"device/child2//","@parent":"device/main//","@type":"child-device"}
     ...    ${entities}
 
+Entity twin update apis
+    ${entity}=    Execute Command
+    ...    curl -X PUT http://localhost:8000/tedge/entity-store/v1/entities/device/main///twin/maintenance_window -H 'Content-Type: application/json' -d '5'
+    ${entity}=    Evaluate    ${entity}
+    Should Be Equal    ${entity["maintenance_window"]}    ${5}
```
Comment (reubenmiller) [inline tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot:134]: When setting a digital twin value, the `PUT` response is different to the `GET` response for the same API endpoint, where it is expected that it should be the same.

For example, below shows the response from adding a new digital twin value:

```
$ tedge http put /tedge/entity-store/v1/entities/device/main///twin/hello --data '"hello"'
{"@topic-id":"device/main//","@type":"device","hello":"hello"}
```

Then retrieving the digital twin value using `GET` returns:

```
$ tedge http get /tedge/entity-store/v1/entities/device/main///twin/hello
"hello"
```

So for this test case I would expect that the user is only interacting with the `twin` data, and not the entity store data.

**Comment by reubenmiller** [comment]
It might be worth change the api endpoint path, as there is some duplication of information which then makes it harder to extend functionality in the future.

**From**
```
/tedge/entity-store/v1/entities/device/main///twin/hello
```

**To**

```sh
/tedge/v1/entities/device/main///twin/hello
```

I also wouldn't be opposed to changing out the word "entities" for a more use-friendly name...but until we have another suggestion, then we will stick with it.

File: crates/core/tedge_agent/src/http_server/entity_store.rs (hunk 1 above)
Comment (reubenmiller) [reply crates/core/tedge_agent/src/http_server/entity_store.rs:266]: We should also prevent empty fragment keys as well as the following leads to an empty key value.

```sh
$ tedge http put '/tedge/entity-store/v1/entities/device/main///twin/' '{"hello":"foo"}'
{"@topic-id":"device/main//","@type":"device","":{"hello":"foo"},"hello":"foo2","hello2":"foo2"}
```

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot (hunk 4 above)
Comment (albinsuresh) [reply tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot:134]: Both `PUT` and `DELETE` are returning the updated entity definition, which I thought the users would benefit from. Of course, they could do another `GET` to get the same. If that isn't desired, then what should the response be? Just `200 OK` is fine?

File: crates/core/tedge_agent/src/http_server/entity_store.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (reubenmiller) [reply crates/core/tedge_agent/src/http_server/entity_store.rs:158]: I'm also having problems reconciling the replacing of all of the existing digital twin keys using the following command:

```sh
$ tedge http put /tedge/entity-store/v1/entities/device/main// '{"foo":"bar"}'
Error: failed to PUT http://127.0.0.1:8000/tedge/entity-store/v1/entities/device/main//

Caused by:
    HTTP client error: 404 Not Found
    {"error":"Path: device/main// not found"}
```

Using `patch` would work to add/edit values, but not to replace all other values.

[the remark this answers is not in the document]
Comment (reubenmiller) [reply crates/core/tedge_agent/src/http_server/entity_store.rs:158]: Or should we keep the entity registration information separate to the digital twin information, and instead it would be a put to:

```sh
$ tedge http put /tedge/entity-store/v1/entities/device/main///twin '{"foo":"bar"}'

# or 
$ tedge http put /tedge/entity-store/v1/entities/device/main///twin/ '{"foo":"bar"}'
```

**Comment by albinsuresh** [comment]
> It might be worth change the api endpoint path, as there is some duplication of information which then makes it harder to extend functionality in the future.
> 
> **From**
> 
> ```
> /tedge/entity-store/v1/entities/device/main///twin/hello
> ```
> 
> **To**
> 
> ```shell
> /tedge/v1/entities/device/main///twin/hello
> ```
> 
> I also wouldn't be opposed to changing out the word "entities" for a more use-friendly name...but until we have another suggestion, then we will stick with it.

Agree about the duplication. The `/tedge/entity-store` prefix was added just to keep it consistent with the `/tedge/file-transfer` endpoints which enabled a clear separation of those service endpoints as well. But, I feel we can achieve the same by using `/v1/entities` prefix as well.

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot [hunk 5]
```diff
@@ -1,5 +1,6 @@
 *** Settings ***
 Resource            ../../../resources/common.resource
+Library             Collections
```
Comment (reubenmiller) [inline tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot:3]: Is this line actually needed? I ran the test without it and it passed.

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot (hunk 4 above)
Comment (reubenmiller) [inline tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot:134]: Robotframework has a keyword to do this rather than using the expression evaluation, `${5}`.

```suggestion
    Should Be Equal As Integers    ${entity["maintenance_window"]}    5
```

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot [hunk 6]
```diff
@@ -126,6 +127,28 @@ Delete entity tree
...
+    Should Be Equal    ${get}    5
+
+    ${timestamp}=    Get Unix Timestamp
+    ${entity}=    Execute Command
+    ...    curl -X DELETE http://localhost:8000/tedge/entity-store/v1/entities/device/main///twin/maintenance_window
+    ${entity}=    Evaluate    ${entity}
+    Should Not Contain    ${entity}    maintenance_window
+    Should Have MQTT Messages
+    ...    te/device/main///twin/maintenance_window
+    ...    date_from=${timestamp}
```
Comment (reubenmiller) [inline tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot:150]: It would be better to check if the retained message has been cleared rather than just checking if any message has been received (as it could also be a value update).

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot (hunk 4 above)
Comment (reubenmiller) [reply tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot:134]: I think the main point is that the `twin` api endpoint is solely about the twin data, and not the entity data. So the `PUT` response should be exactly the same as the `GET` response and both return a HTTP 200 status code.

For the `DELETE` twin command, I would just return a `204` (No Content). This is different to the `DELETE` on the entity store, as the entity store was deleting a tree of devices, so it was unclear which devices are being deleted (as it behaved more like a query), where as for deleting a digital twin property, the context is already clear in the URL, and the user does not care about the value (that's why they are deleting it).

File: crates/core/tedge_agent/src/http_server/entity_store.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/core/tedge_agent/src/http_server/entity_store.rs]: Resolved by f837935c

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot (hunk 4 above)
Comment (albinsuresh) [reply tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot]: Resolved by f837935c

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot (hunk 5 above)
Comment (albinsuresh) [reply tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot]: Resolved by f837935c

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot (hunk 4 above)
Comment (albinsuresh) [reply tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot]: Resolved by f837935c

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot (hunk 6 above)
Comment (albinsuresh) [reply tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot:147]: Resolved by f837935c

File: crates/core/tedge_agent/src/http_server/entity_store.rs (hunk 1 above)
Comment (didier-wenzek) [reply crates/core/tedge_agent/src/http_server/entity_store.rs]: > Resolved by f837935c

Now, working as expected

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot (hunk 4 above)
Comment (didier-wenzek) [reply tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot]: > Resolved by f837935

Perfect. Now, `PUT` returns no content and `GET` returns what has been `PUT`. `DELETE` returns `204 No Content` as expected.

File: crates/core/tedge_agent/src/http_server/entity_store.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/core/tedge_agent/src/http_server/entity_store.rs:162]: >> But the variant above gives an error message that is difficult to understand

> Yeah, that's misleading. It should have been a 405 Method Not Allowed.

Okay now, the following returns a `405 Method Not Allowed`.

```
$ tedge http patch /tedge/entity-store/v1/entities/device/main///twin/maintenance_window '{"days":43}'
Error: failed to PATCH http://127.0.0.1:8000/tedge/entity-store/v1/entities/device/main///twin/maintenance_window

Caused by:
    HTTP client error: 405 Method Not Allowed
    {"error":"Method Not Allowed"}
```

However it would be a bit better to tell what is actually not allowed: i.e. to patch a fragment.

Also, as patching a fragment is not allowed it would be better to return a `405 Method Not Allowed` in the following case:

```
$ tedge http patch /tedge/entity-store/v1/entities/device/main///twin/maintenance_window '"42"'
Error: failed to PATCH http://127.0.0.1:8000/tedge/entity-store/v1/entities/device/main///twin/maintenance_window

Caused by:
    HTTP client error: 422 Unprocessable Entity
    Failed to deserialize the JSON body into the target type: invalid type: string "42", expected a map at line 1 column 4
```

**Review by didier-wenzek — COMMENTED** [review]
I will be happy to approve - once the unit tests fixed. Also, I have some suggestions to improve the error messages.

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot (hunk 4 above)
Comment (reubenmiller) [reply tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot]: Something is still a bit strange. Doing a PUT does return something (which is expected imo), however it only returns something when the value has changed (which is unexpected).

For example, the following shows the unexpected behaviour:

```sh
# first time returns the value which was 'put', 5
$ tedge http put /tedge/entity-store/v1/entities/device/main///twin/maintenance_window '5'

# second time, no response is given
$ tedge http put /tedge/entity-store/v1/entities/device/main///twin/maintenance_window '5'
```

Let's keep the response stateless, and just always return the current value (yes this may be the same as what was sent), but it is there to confirm that the server's interpretation of the payload is in fact the same.

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot [hunk 7]
```diff
@@ -126,6 +113,134 @@ Delete entity tree
...
+
+    ${timestamp}=    Get Unix Timestamp
+    ${http_code}=    Execute Command
+    ...    curl -o /dev/null --silent --write-out "%\{http_code\}" -X DELETE http://localhost:8000/tedge/entity-store/v1/entities/device/main///twin/maintenance_window
+    Should Be Equal    ${http_code}    204
+    Should Have MQTT Messages
+    ...    te/device/main///twin/maintenance_window
+    ...    date_from=${timestamp}
+    ${retained_message}=    Execute Command
+    ...    tedge mqtt sub --no-topic te/device/main///twin/maintenance_window --duration 1
```
Comment (reubenmiller) [inline tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot:136]: Yes the duration will default to seconds, but it would be better to be explicit, as the default unit is not clear to the reader.

```suggestion
    ...    tedge mqtt sub --no-topic te/device/main///twin/maintenance_window --duration 1s
```

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot [hunk 8]
```diff
@@ -126,6 +113,134 @@ Delete entity tree
...
+    ${new_payload}=    Set Variable    {"maintainer":"John Doe","maintenance_mode":true,"maintenance_window":5}
+    ${put}=    Execute Command
+    ...    curl -X PUT http://localhost:8000/tedge/entity-store/v1/entities/device/main///twin -H 'Content-Type: application/json' -d '${new_payload}'
+    Should Be Equal    ${put}    ${new_payload}
+    Should Have MQTT Messages
+    ...    te/device/main///twin/maintenance_mode
+    ...    message_contains=true
+    Should Have MQTT Messages
+    ...    te/device/main///twin/maintenance_window
+    ...    message_contains=5
```
Comment (reubenmiller) [inline tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot:151]: It would be worth while also adding a test for replacing all of the existing values with an empty object, as currently there is a bit of an inconsistency when putting `{}` to the `/twin/` endpoint.

For example:

```sh
tedge http put /tedge/entity-store/v1/entities/device/main///twin/ '{}'

# Current: <no response>
# Expected: {}
```

**Comment by reubenmiller** [comment]
It's shaping up nicely. I just found some slightly unexpected results whilst testing.

### Unexpected json payloads which are accepted

This is probably due to the crate we're using to parse the JSON, but I found some unexpected results when sending invalid json payloads (I wouldn't say that I'm against it, but wanted to point it out as it might lead to some unexpected results for users). I'm assuming the crate is doing some json token parsing and only parsing the first "token" (e.g. integer, string, object, array etc.)

All of the following digital twin updates work and don't return an error (despite the payloads technically not being full valid json (from start to finish))

```sh
$ tedge http put /tedge/entity-store/v1/entities/device/main///twin/test '12-2'
=> 12

$ tedge http put /tedge/entity-store/v1/entities/device/main///twin/test '12.12.1'
=> 12.12

$ tedge http put /tedge/entity-store/v1/entities/device/main///twin/test 'true true'
=> true

$ tedge http put /tedge/entity-store/v1/entities/device/main///twin/test '{"a":1}{"b":2}'
=> {"a":1}
```


### What size limit to we impose on the payloads?

This would be useful to users to at least guide them to not publishing huge payloads, and we should make sure it aligns with the max payload size used by the local MQTT broker (if we have some limit set).

Below shows the current response when trying to upload a 5MB json content where the request is rejected (which is good), however I suspect it wasn't an active check, and just the default buffer size that we're using.

```sh
$ ls -lh body.json
-rw-r--r-- 1 root root 5.7M Mar 20 18:39 body.json

$ curl -X PUT http://localhost:8000/tedge/entity-store/v1/entities/device/main///twin/maintenance_window -d @body.json -H "Content-Type: application/json"
Failed to buffer the request body: length limit exceeded
```

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot [hunk 9]
```diff
@@ -126,6 +108,138 @@ Delete entity tree
...
+    ${get}=    Execute Command
+    ...    curl http://localhost:8000/tedge/entity-store/v1/entities/device/main///twin
+    Should Be Equal
+    ...    ${get}
+    ...    ${new_payload}
+
+    # Replace existing twins
+    ${timestamp}=    Get Unix Timestamp
+    ${new_payload}=    Set Variable    {"maintenance_mode":false}
+    ${put}=    Execute Command
```
Comment (reubenmiller) [inline tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot:157]: This looks to be the offending line which is causing the RF lint error (e.g. `${put}` assigned to be never used.

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot (hunk 8 above)
Comment (albinsuresh) [reply tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot]: Resolved by 32305a5

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot (hunk 9 above)
Comment (albinsuresh) [reply tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot:167]: Resolved by 7d725d7

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot (hunk 7 above)
Comment (albinsuresh) [reply tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot]: Resolved by 6156bcb

File: tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot (hunk 4 above)
Comment (albinsuresh) [reply tests/RobotFramework/tests/tedge_agent/http_server/entity_store_api.robot]: Resolved by 32305a5. I was trying to give a more accurate response to the user whether the update really updated something or not. But, since keeping these methods as idempotent is the convention/desired, I've changed it.

**Review by didier-wenzek — APPROVED** [review]
Approved. Thank you for your perseverance polishing this new API.

I would just consider to simplify the api endpoint prefix as suggested https://github.com/thin-edge/thin-edge.io/pull/3474#issuecomment-2733462317

**Comment by albinsuresh** [comment]
> I would just to consider to simplify the api endpoint prefix as suggested [#3474 (comment)](https://github.com/thin-edge/thin-edge.io/pull/3474#issuecomment-2733462317)

Yup, that's the plan. Had discussed it with @reubenmiller offline and felt that it would be better to do it as an independent PR.

**Comment by reubenmiller** [comment]
> > ### What size limit to we impose on the payloads?
> > This would be useful to users to at least guide them to not publishing huge payloads, and we should make sure it aligns with the max payload size used by the local MQTT broker (if we have some limit set).
> > Below shows the current response when trying to upload a 5MB json content where the request is rejected (which is good), however I suspect it wasn't an active check, and just the default buffer size that we're using.
> > ```shell
> > $ ls -lh body.json
> > -rw-r--r-- 1 root root 5.7M Mar 20 18:39 body.json
> > 
> > $ curl -X PUT http://localhost:8000/tedge/entity-store/v1/entities/device/main///twin/maintenance_window -d @body.json -H "Content-Type: application/json"
> > Failed to buffer the request body: length limit exceeded
> > ```
> 
> Resolved by [989c944](https://github.com/thin-edge/thin-edge.io/commit/989c944b665bef7cf9066b67a101a7fb421593c0).

This doesn't seem to be working as expected...I get a buffer size error for a file size around 6.6MB before it hits the upper limit of the defined threshold)

```sh
curl -X PUT http://localhost:8000/tedge/entity-store/v1/entities/device/main///twin -d @body2.json -H "Content-Type: application/json"
Failed to buffer the request body: length limit exceeded
```

It seems that we need to set the max body size on the [axum server](https://docs.rs/axum/latest/axum/extract/struct.DefaultBodyLimit.html).

File: crates/core/tedge_agent/src/http_server/entity_store.rs [hunk 10]
```diff
@@ -170,35 +188,141 @@ async fn register_entity(
...
+async fn put_resource(
+    State(state): State<AgentState>,
+    Path(path): Path<String>,
+    payload: String,
+) -> impl IntoResponse {
+    if payload.len() > HTTP_MAX_PAYLOAD_SIZE {
+        return Err(Error::EntityStoreError(
+            entity_store::Error::PayloadSizeExceeded(payload.len()),
+        ));
+    }
```
Comment (albinsuresh) [inline crates/core/tedge_agent/src/http_server/entity_store.rs:224]: Wasted a lot of time trying to use the [RequestBodyLimitLayer](https://docs.rs/tower-http/0.6.2/tower_http/limit/struct.RequestBodyLimitLayer.html) with axum `Router`, but failed. So, falling back to this instead.

Comment (albinsuresh) [reply crates/core/tedge_agent/src/http_server/entity_store.rs]: Well... It was my bad. I was wrongly trying to integrate the `tower-http` abstraction [RequestBodyLimitLayer](https://docs.rs/tower-http/0.6.2/tower_http/limit/struct.RequestBodyLimitLayer.html) instead [DefaultBodyLimit](https://docs.rs/axum/latest/axum/extract/struct.DefaultBodyLimit.html) which is the right `axum` router level abstraction. Resolved by 69a6f6f.

**Comment by albinsuresh** [comment]
> This doesn't seem to be working as expected...I get a buffer size error for a file size around 6.6MB before it hits the upper limit of the defined threshold)
> 
> ```shell
> curl -X PUT http://localhost:8000/tedge/entity-store/v1/entities/device/main///twin -d @body2.json -H "Content-Type: application/json"
> Failed to buffer the request body: length limit exceeded
> ```
> 
> It seems that we need to set the max body size on the [axum server](https://docs.rs/axum/latest/axum/extract/struct.DefaultBodyLimit.html).

Resolved in 69a6f6f by applying the 1 MB limit as agreed.

**Review by reubenmiller — APPROVED** [review]
Approved. Thanks for the persistence!
